File: src/main/java/com/ptmanager/backend/swap/SwapRequestController.java
package com.ptmanager.backend.swap;

import java.time.LocalDate;
import java.util.List;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.ptmanager.backend.domain.SwapRequest;
import com.ptmanager.backend.domain.SwapRequestStatus;

@RestController
@RequestMapping("/api/swap-requests")
public class SwapRequestController {

    private final SwapRequestService swapRequestService;

    public SwapRequestController(SwapRequestService swapRequestService) {
        this.swapRequestService = swapRequestService;
    }

    @GetMapping
    public List<SwapRequest> findSwapRequests() {
        return swapRequestService.findSwapRequests();
    }

    @PostMapping
    public SwapRequest createSwapRequest(@Valid @RequestBody CreateSwapRequest request) {
        return swapRequestService.createSwapRequest(
            request.requesterId(),
            request.substituteId(),
            request.workDate(),
            request.reason()
        );
    }

    @PatchMapping("/{id}/status")
    public SwapRequest updateStatus(
        @PathVariable long id,
        @Valid @RequestBody UpdateSwapRequestStatus request
    ) {
        return swapRequestService.updateStatus(id, request.status());
    }

    public record CreateSwapRequest(
        @NotNull Long requesterId,
        Long substituteId,
        @NotNull @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate workDate,
        @NotBlank String reason
    ) {
    }

    public record UpdateSwapRequestStatus(
        @NotNull SwapRequestStatus status
    ) {
    }
}

File: src/main/kotlin/com/ptmanager/backend/swaprequest/SwapRequestController.kt
package com.ptmanager.backend.swaprequest

import com.ptmanager.backend.domain.SwapRequest
import com.ptmanager.backend.domain.SwapRequestStatus
import com.ptmanager.backend.swaprequest.dto.ApproveSwapRequest
import com.ptmanager.backend.swaprequest.dto.CreateSwapRequest
import com.ptmanager.backend.swaprequest.dto.SwapApplicationResponse
import com.ptmanager.backend.swaprequest.dto.SwapRequestDetail
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/swap-requests")
class SwapRequestController(
    private val swapRequestService: SwapRequestService,
) {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun createSwapRequest(
        @AuthenticationPrincipal userId: Long,
        @Valid @RequestBody request: CreateSwapRequest,
    ): SwapRequest = swapRequestService.createSwapRequest(userId, request.shiftId, request.reason)

    @GetMapping
    fun findSwapRequests(
        @AuthenticationPrincipal userId: Long,
        @RequestParam workplaceId: Long,
        @RequestParam view: String,
        @RequestParam(required = false) status: SwapRequestStatus?,
    ): List<SwapRequest> = swapRequestService.findSwapRequests(workplaceId, view, userId, status)

    @GetMapping("/{swapRequestId}")
    fun getDetail(@PathVariable swapRequestId: Long): SwapRequestDetail =
        swapRequestService.getDetail(swapRequestId)

    @PostMapping("/{swapRequestId}/applications")
    @ResponseStatus(HttpStatus.CREATED)
    fun apply(
        @AuthenticationPrincipal userId: Long,
        @PathVariable swapRequestId: Long,
    ): SwapApplicationResponse = swapRequestService.apply(swapRequestId, userId)

    @GetMapping("/{swapRequestId}/applications")
    @PreAuthorize("hasRole('EMPLOYER')")
    fun listApplications(@PathVariable swapRequestId: Long): List<SwapApplicationResponse> =
        swapRequestService.listApplications(swapRequestId)

    @PostMapping("/{swapRequestId}/approve")
    @PreAuthorize("hasRole('EMPLOYER')")
    fun approve(
        @PathVariable swapRequestId: Long,
        @Valid @RequestBody request: ApproveSwapRequest,
    ): SwapRequest = swapRequestService.approve(swapRequestId, request.applicantId)

    @PostMapping("/{swapRequestId}/reject")
    @PreAuthorize("hasRole('EMPLOYER')")
    fun reject(@PathVariable swapRequestId: Long): SwapRequest = swapRequestService.reject(swapRequestId)
}
